File: backend/app/schemas/inventory.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class InventoryBase(BaseModel):
    """
    Shared properties for inventory models. Visible by anyone.
    """

    area: str = Field(..., description="The area of the inventory item.")
    gauge: str = Field(..., description="The gauge of the inventory item.")
    mill_no: str = Field(..., description="The mill number of the inventory item.")
    nwt_lbs: float = Field(..., description="The NWT in pounds of the inventory item.")
    osf_no: str = Field(..., description="The OSF number of the inventory item.")
    position: str = Field(..., description="The position of the inventory item.")
    receipt_date: datetime = Field(
        ..., description="The receipt date of the inventory item."
    )
    row: str = Field(..., description="The row of the inventory item.")
    supplier_uuid: UUID = Field(
        ..., description="The supplier associated with the inventory item."
    )
    status: str = Field(..., description="The status of the inventory item.")
    width_inches: str = Field(
        ..., description="The width in inches of the inventory item."
    )
    bay: str = Field(..., description="The bay of the inventory item.")
    bl: str = Field(..., description="The BL of the inventory item.")
    pl: str = Field(..., description="The PL of the inventory item.")
    po: str = Field(..., description="The PO of the inventory item.")


class InventoryCreate(InventoryBase):
    """
    Properties to create a new inventory item. Visible by anyone.
    """

    pass


class InventoryUpdate(BaseModel):
    """
    Properties to update an inventory item. Visible by anyone.
    """

    area: str | None = Field(default=None, description="The area of the inventory item.")
    gauge: str | None = Field(
        default=None, description="The gauge of the inventory item."
    )
    mill_no: str | None = Field(
        default=None, description="The mill number of the inventory item."
    )
    nwt_lbs: float | None = Field(
        default=None, description="The NWT in pounds of the inventory item."
    )
    osf_no: str | None = Field(
        default=None, description="The OSF number of the inventory item."
    )
    position: str | None = Field(
        default=None, description="The position of the inventory item."
    )
    receipt_date: datetime | None = Field(
        default=None, description="The receipt date of the inventory item."
    )
    row: str | None = Field(default=None, description="The row of the inventory item.")
    supplier_uuid: UUID | None = Field(
        default=None, description="The supplier associated with the inventory item."
    )
    status: str | None = Field(
        default=None, description="The status of the inventory item."
    )
    width_inches: str | None = Field(
        default=None, description="The width in inches of the inventory item."
    )
    bay: str | None = Field(default=None, description="The bay of the inventory item.")
    bl: str | None = Field(default=None, description="The BL of the inventory item.")
    pl: str | None = Field(default=None, description="The PL of the inventory item.")
    po: str | None = Field(default=None, description="The PO of the inventory item.")


class Inventory(InventoryBase):
    """
    Properties shared by models stored in the database. Visible by anyone.
    """

    uuid: UUID = Field(..., description="Unique identifier for the inventory item.")
    last_updated: datetime = Field(
        ..., description="Last updated timestamp for the inventory item."
    )

File: frontend/src/models/inventory.ts
import { Form } from './form'
import { Column } from './column'

export interface Inventory {
  uuid: string
  area: string
  gauge: string
  mill_no: string
  nwt_lbs: number
  osf_no: string
  position: string
  receipt_date: string
  row: string
  supplier_uuid: string
  status: string
  width_inches: string
  bay: string
  bi: string
  pl: string
  po: string
  last_updated: string
}

export const InventoryForm: Form = {
  title: 'Inventory Information',
  groups: [
    {
      title: 'Supplier Information',
      fields: [
        {
          name: 'supplier_uuid',
          label: 'Supplier UUID',
          type: 'text',
          required: true,
          disabled: true,
        },
      ],
    },
    {
      title: 'Coil Information',
      fields: [
        //{ name: 'manufacturing', label: 'Manufacturing', type: 'text' },
        { name: 'mill_no', label: 'Mill No', type: 'text', required: true },
        { name: 'nwt_lbs', label: 'NWT (lbs)', type: 'number', required: true },
        { name: 'gauge', label: 'Gauge', type: 'text', required: true },
        { name: 'width_inches', label: 'Width (inches)', type: 'text', required: true },
        { name: 'osf_no', label: 'OSF No', type: 'text', required: true },
        {
          name: 'status',
          label: 'Status',
          type: 'select',
          required: true,
          options: [
            { label: 'Queued', value: 'queued' },
            { label: 'In Procurement', value: 'in_procurement' },
            { label: 'Ordered', value: 'ordered' },
            { label: 'Received', value: 'received' },
            { label: 'In Inventory', value: 'in_inventory' },
            { label: 'Reserved', value: 'reserved' },
            { label: 'Checked Out', value: 'checked_out' },
            { label: 'In Production', value: 'in_production' },
            { label: 'Depleted', value: 'depleted' },
            { label: 'Disposed', value: 'disposed' },
            { label: 'Cancelled', value: 'cancelled' },
          ],
        },
      ],
    },
    {
      title: 'Storing Information',
      fields: [
        { name: 'area', label: 'Area', type: 'text', required: true },
        { name: 'bay', label: 'Bay', type: 'text', required: true },
        { name: 'row', label: 'Row', type: 'text', required: true },
        { name: 'position', label: 'Position', type: 'text', required: true },
      ],
    },
    {
      title: 'Purchase Information',
      fields: [{ name: 'po', label: 'PO', type: 'text', required: true }],
    },
    {
      title: 'Shipment Information',
      fields: [
        { name: 'bi', label: 'BI', type: 'text', required: true },
        { name: 'pl', label: 'PL', type: 'text', required: true },
      ],
    },
    {
      title: 'Receiving Information',
      fields: [{ name: 'receipt_date', label: 'Receipt Date', type: 'date', required: true }],
    },
    // {
    //   title: 'QC Information',
    //   fields: [{ name: 'inspected', label: 'Inspected', type: 'text' }],
    // },
  ],
}

export const InventoryColumns: Column<Inventory>[] = [
  { key: 'area' as keyof Inventory, label: 'Area' },
  { key: 'gauge' as keyof Inventory, label: 'Gauge' },
  { key: 'mill_no' as keyof Inventory, label: 'Mill No' },
  { key: 'nwt_lbs' as keyof Inventory, label: 'NWT (lbs)' },
  { key: 'osf_no' as keyof Inventory, label: 'OSF No' },
  { key: 'position' as keyof Inventory, label: 'Position' },
  { key: 'receipt_date' as keyof Inventory, label: 'Receipt Date' },
  { key: 'row' as keyof Inventory, label: 'Row' },
  { key: 'supplier_uuid' as keyof Inventory, label: 'Supplier UUID' },
  { key: 'status' as keyof Inventory, label: 'Status' },
  { key: 'width_inches' as keyof Inventory, label: 'Width (inches)' },
  { key: 'last_updated' as keyof Inventory, label: 'Last Updated' },
]
